### ISSUE TITLE
Embind function not runs syncronously. Using Asyncify.

### ISSUE THREAD

**Issue by Satancito** [emscripten-core/emscripten#10850]
I need to do a HTTP request to server, the result is JSON. The result is used in another C++ function which is exported using embind exports. 

When call this exported function from js, it runs fast and cannot store value from exported function, value is empty, later in 2 seconds function is completed but value never was assigned.

**Steps to reproduce**

**1. My test/fake webapi**
[https://jsonplaceholder.typicode.com/](https://jsonplaceholder.typicode.com/)

<hr />

**2. Mi JS Code - Instance.js**

```
window.VModule({ noInitialRun: false }).then(module => {
     window.VMain = {
          Module: module
     };
});

window.GetData = async function GetData() {
     try {
          var url = "https://jsonplaceholder.typicode.com/todos/200"
          var options = {
               method: "GET",
               //mode: "cors",
               cache: "default"
          };
          var response = await fetch(new Request(url), options);
          var json = await response.json();
          console.log("DATA IN JS: " + json.title);
          return json.title;
     }
     catch (error) {
          console.log("DATA IN JS: ERROR_JS");
          return "ERROR_JS";
     }
}
```
<hr />

**3. My C++ code - main.cpp** 
- EM_JS_GetData function who calls "window.GetData" defined in Instance.js
- GetData() calls EM_JS_GetData
window.GetData() -> EM_JS_GetData -> GetData() -> Button call from index.html.

```
#include <iostream>
#include <string>
#include <emscripten.h>
#include <emscripten/bind.h>
typedef std::string String;

EM_JS(const char *, EM_JS_GetData, (), {
    return Asyncify.handleSleep(function(wakeUp) {   
        window.GetData()
        .then(data=>{
            //Converting JsString to C-String
            var lengthBytes = lengthBytesUTF8(data)+1;
            var result = _malloc(lengthBytes);
            stringToUTF8(data, result, lengthBytes);
            wakeUp(result);
        })
    });
});

String GetData()
{
    try
    {
        const char *cstr = EM_JS_GetData();
        String result(cstr);
        free(static_cast<void*> (const_cast<char *>(cstr)));
        emscripten_console_log((u8"DATA IN C++: " + result).c_str());
        return result;
    }
    catch (...)
    {
        emscripten_console_log(u8"DATA IN CPP: ERROR_CPP");
        return "ERROR_CPP";
    }
}


int main(int argc, char ** argv) 
{
    std::cout<< "All components initialized" << std::endl;       
}

EMSCRIPTEN_BINDINGS(my_module) 
{
    using namespace emscripten;
    function("GetData", &GetData);
}
```
<hr />

**4. My compile command in power shell core syntax. - Compile.ps1**

```
em++.bat `
"main.cpp" `
-o "VModule.js" `
-std=c++17 `
-O2 `
--bind `
-s WASM=1 `
-s SINGLE_FILE=1 `
-s ASYNCIFY=1 `
-s NO_EXIT_RUNTIME=1 `
-s EXTRA_EXPORTED_RUNTIME_METHODS=[`'ccall`',`'cwrap`'] `
-s EXPORT_NAME=`'VModule`' `
-s MODULARIZE=1 `
-s DISABLE_EXCEPTION_CATCHING=0 `
-s ENVIRONMENT=web `
-s ASYNCIFY_IMPORTS=[`'EM_JS_GetData`'] `
-s VERBOSE=0 `
```
<hr />

**5. Index.html**

```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Validatos - WebAssembly</title>
    <link rel="icon" href="Assets/favicon/favicon.ico">
    <script type="text/javascript" src="VModule.js"></script>
    <script type="module" src="Instance.js"></script>
</head>

<body>
    <button onclick="OnBtnClick()">Test</button>
    <script type="text/javascript">
        function OnBtnClick() {
            console.log("BEFORE");
            var title = window.VMain.Module.GetData();//RESULT FROM FUNCTION IS NO ASSIGNED.
            console.log("DATA IN INDEX.HTML: " + title);//console prints empty string.
            console.log("AFTER");
        }
    </script>
</body>
</html>
```

When I clicked the button, I get following result.
```
All components initialized
(index):17 BEFORE
(index):19 DATA IN INDEX.HTML: 
(index):20 AFTER
Instance.js:19 DATA IN JS: ipsam aperiam voluptates qui
VModule.js:9 DATA IN C++: ipsam aperiam voluptates qui
```
![image](https://user-images.githubusercontent.com/18097869/78508552-427e9300-774d-11ea-92c7-664f0ecc57d1.png)

how can I call syncronously window.VMain.Module.GetData()?
Desired behaviour

```
All components initialized
(index):17 BEFORE
Instance.js:19 DATA IN JS: ipsam aperiam voluptates qui
VModule.js:9 DATA IN C++: ipsam aperiam voluptates qui
(index):19 DATA IN INDEX.HTML: ipsam aperiam voluptates qui
(index):20 AFTER
```

**Comment by kripken** [comment]
There are a lot of difference pieces here: asyncify, embind, C++ exceptions, EM_JS, async/await in JS - it's possible some combination of those is hitting a bug. To investigate I'd recommend simplifying the testcase further to remove even more - for example, remove the C++ exceptions handling code and see if the problem still happens, see if replacing the async/await in JS with something simpler (like a setTimeout) works, and so forth.

**Comment by JomaCorpFX** [comment]
New Changes. The problem persists.
**1. main.cpp**
Using this examples
[Making async Web APIs behave as if they were synchronous](https://emscripten.org/docs/porting/asyncify.html#making-async-web-apis-behave-as-if-they-were-synchronous)
[Returning values](https://emscripten.org/docs/porting/asyncify.html#returning-values)

```
EM_JS(const char *, EM_JS_GetData, (), {
    return Asyncify.handleSleep(function(wakeUp) {   
        var url = "https://jsonplaceholder.typicode.com/todos/200";
          var options = {
               method: "GET",
               //mode: "cors",
               cache: "default"
          };
          fetch(new Request(url), options)
          .then(response => {
              console.log("RESPONSE OK");
              return response.json();
          })
          .then(json=>{
              var data = String(json.title);
              console.log("DATA IN JS: " + data);
              var lengthBytes = lengthBytesUTF8(data)+1;
              var result = _malloc(lengthBytes);
              stringToUTF8(data, result, lengthBytes);
              setTimeout(wakeUp(result),0);
          })
          .catch(error=>{ //TESTED WITHOUT THIS BLOCK /SAME RESULT
              var data = "ERROR-JS";
              var lengthBytes = lengthBytesUTF8(data)+1;
              var result = _malloc(lengthBytes);
              stringToUTF8(data, result, lengthBytes);
              setTimeout(wakeUp(result),0);
          });
    });
});

String GetData()
{
        //TESTED WITH TRY CATCH BLOCK / SAME RESULT
        emscripten_console_log("CPP START");
        const char *cstr = EM_JS_GetData();
        emscripten_console_log("EM_JS ENDED");
        String result(cstr);
        free(static_cast<void*> (const_cast<char *>(cstr)));
        emscripten_console_log((u8"DATA IN C++: " + result).c_str());
        return result;
}

int main(int argc, char ** argv) 
{
    std::cout<< "All components initialized" << std::endl;       
}
```

<hr />

**2. Instance.js**

```
window.VModule({ noInitialRun: false }).then(module => {
     window.VMain = {
          Module: module
     };
});
```
<hr />

**3. Index.html**

```
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>VModuleTest - WebAssembly</title>
    <link rel="icon" href="Assets/favicon/favicon.ico">
    <script type="text/javascript" src="VModule.js"></script>
    <script type="module" src="Instance.js"></script>
</head>

<body>
    <button onclick="OnBtnClick()">Test</button>
    <script type="text/javascript">
        async function OnBtnClick() {
            console.log("1. BEFORE");
            var result = window.VMain.Module.GetData();//Result from exported C++ function is never assigned. Empty string "" is assigned. Function not runs syncronously here.
            console.log("2. DATA IN INDEX.HTML: " + result);//console prints empty string.
            console.log("3. AFTER");
            //output from EM_JS function and C++ function is printed in console. TOO LATE.
        }
    </script>
</body>

</html>
```
<hr />

**Output when button is clicked**
![image](https://user-images.githubusercontent.com/18097869/78853893-a15e3980-79e5-11ea-8d1d-61c22e1fe716.png)

**DESIRED BEHAVIOUR**
```
All components initialized
(index):24 1. BEFORE
VModule.js:9 CPP START
VModule.js:9 RESPONSE OK
VModule.js:9 DATA IN JS: ipsam aperiam voluptates qui
VModule.js:9 EM_JS ENDED
VModule.js:9 DATA IN C++: ipsam aperiam voluptates qui
(index):26 2. DATA IN INDEX.HTML: ipsam aperiam voluptates qui ------> Correct value is printed
(index):27 3. AFTER
```

**Comment by JomaCorpFX** [comment]
Update.

Adding exported function with C Linkage using extern "C" - GetDataEx
When call from browser with ccall, this function returns a promise instead of a string. This promise is a resolved promise with the value.
With this can be resolve my problem but, is this the correct behaviour when call a function with CCALL?
C Function returns char* but in JS we get a promise(with the result string inside) instead of string.

**1. main.cpp**
```
EM_JS(const char *, EM_JS_GetData, (), {
    return Asyncify.handleSleep(function(wakeUp) {   
        var url = "https://jsonplaceholder.typicode.com/todos/200";
          var options = {
               method: "GET",
               //mode: "cors",
               cache: "default"
          };
          fetch(new Request(url), options)
          .then(response => {
              console.log("RESPONSE OK");
              return response.json();
          })
          .then(json=>{
              var data = String(json.title);
              console.log("DATA IN JS: " + data);
              var lengthBytes = lengthBytesUTF8(data)+1;
              var result = _malloc(lengthBytes);
              stringToUTF8(data, result, lengthBytes);
              setTimeout(wakeUp(result),0); 
          })
          .catch(error=>{ 
              var data = "ERROR-JS";
              var lengthBytes = lengthBytesUTF8(data)+1;
              var result = _malloc(lengthBytes);
              stringToUTF8(data, result, lengthBytes);
              setTimeout(wakeUp(result),0);
          });
    });
});

String GetData()
{
        //TESTED WITH TRY CATCH BLOCK / SAME RESULT
        emscripten_console_log("CPP START");
        const char *cstr = EM_JS_GetData();
        emscripten_console_log("EM_JS ENDED");
        String result(cstr);
        free(static_cast<void*> (const_cast<char *>(cstr)));
        emscripten_console_log((u8"DATA IN C++: " + result).c_str());
        return result;
}

extern "C" const char * EMSCRIPTEN_KEEPALIVE GetDataEx()
{
    return GetData().c_str();
}


int main(int argc, char ** argv) 
{
    std::cout<< "All components initialized" << std::endl;       
}
```

<hr />

**2. Instance.js**

```
window.VModule({ noInitialRun: false }).then(module => {
     window.VMain = {
          Module: module
     };
});
```
<hr />

**3. Compile.ps1**
added the imported functions param

```
-s EXPORTED_FUNCTIONS=[`'_main`',`'_GetDataEx`'] `
```

<hr />

**4. Index.html**

```
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>VModuleTest - WebAssembly</title>
    <link rel="icon" href="Assets/favicon/favicon.ico">
    <script type="text/javascript" src="VModule.js"></script>
    <script type="module" src="Instance.js"></script>
</head>

<body>
    <button onclick="OnBtnClick()">Test</button>
    <script type="text/javascript">
        async function OnBtnClick() {
            console.log("1. BEFORE");
            var result = window.VMain.Module.ccall('GetDataEx', 'string', [], []); // This function return a resolved promise. // return type is string but returns a promise.
            result.then(valueFromCPP => {
                //Extrange result but resolves the problem. I think this is a bug, or is the correct behaviour.
                console.log("DATA IN INDEX.HTML from PROMISE" + valueFromCPP);
            });
            console.log("2. DATA IN INDEX.HTML: " + result);//console prints empty string.
            console.log("3. AFTER");
    </script>
</body>

</html>
```
<hr />

**Output when button is clicked**
CORRECT RESULT. But I have doubts if this operation is correct.

```
All components initialized
(index):17 1. BEFORE
VModule.js:9 CPP START
(index):23 2. DATA IN INDEX.HTML: [object Promise]
(index):24 3. AFTER
VModule.js:9 RESPONSE OK
VModule.js:9 DATA IN JS: ipsam aperiam voluptates qui
VModule.js:9 EM_JS ENDED
VModule.js:9 DATA IN C++: ipsam aperiam voluptates qui
(index):21 DATA IN INDEX.HTML from PROMISE: ipsam aperiam voluptates qui
```
![image](https://user-images.githubusercontent.com/18097869/78855586-ea17f180-79e9-11ea-8c53-0de95df7a142.png)

**Comment by Satancito** [comment]
Solved with [#11890](https://github.com/emscripten-core/emscripten/issues/11890#issuecomment-711454589)

Check the code
[EmbindAsync](https://github.com/Satancito/EmbindAsync)
